Document AOT analysis warnings (#30967)
dotnet/docs: docs/core/deploying/native-aot/warnings/il3055.md

---

> title: "IL3055: P/invoke method declares a parameter with an abstract delegate"

Reviewer: ```suggestion
title: "IL3055: P/Invoke method declares a parameter with an abstract delegate"
```

> description: "Correctness of interop for abstract delegates cannot be guaranteed after native compilation - the marshalling code for the delegate might not be available"

Reviewer: ```suggestion
description: "Learn about warning IL3055: Correctness of interop for abstract delegates can't be guaranteed after native compilation because the marshalling code for the delegate might not be available."
```

ms.date: 09/01/2022
ms.topic: reference
ms.author: michals
author: MichalStrehovsky
f1_keywords:
  - "IL3055"
---

> # IL3055: P/invoke method declares a parameter with an abstract delegate

Reviewer: ```suggestion
# IL3055: P/Invoke method declares a parameter with an abstract delegate
```

## Cause

> P/invoke marshalling code needs to be generated ahead of time. If marshalling code for a delegate wasn't pregenerated, p/invoke marshalling will fail with an exception at runtime.

Reviewer: ```suggestion
P/Invoke marshalling code needs to be generated ahead of time. If marshalling code for a delegate wasn't pregenerated, P/Invoke marshalling will fail with an exception at run time.
```

Reviewer: Looks like we always capitalize P/Invoke: https://docs.microsoft.com/en-us/dotnet/standard/native-interop/pinvoke

> Marshalling code is generated for delegate types that are used in signatures of p/invoke methods, appear as fields of types passed to native code via p/invoke, or are decorated with <xref:System.Runtime.InteropServices.UnmanagedFunctionPointerAttribute>.

Reviewer: ```suggestion
Marshalling code is generated for delegate types that either:

- Are used in signatures of P/Invoke methods.
- Appear as fields of types passed to native code via P/Invoke.
- Are decorated with <xref:System.Runtime.InteropServices.UnmanagedFunctionPointerAttribute>.
```

> If a concrete type cannot be inferred from the p/invoke signature, marshalling code might not be available at runtime and the p/invoke will throw an exception.

Reviewer: ```suggestion
If a concrete type cannot be inferred from the P/Invoke signature, marshalling code might not be available at run time and the P/Invoke will throw an exception.
```

> Replace <xref:System.Delegate> or <xref:System.MulticastDelegate> in the p/invoke signature with a concrete delegate type.

Reviewer: ```suggestion
Replace <xref:System.Delegate> or <xref:System.MulticastDelegate> in the P/Invoke signature with a concrete delegate type.
```
